### ISSUE TITLE
Huge difference in modelling time (R and Python version) for 75 million observations

### ISSUE THREAD

**Issue by pranavpandya84** [Microsoft/LightGBM#1291]
This is related to currently active machine competition on Kaggle namely TalkingData. In an experimental run, I have tested R version of LightGBM and have compared it with Python version. Surprisingly, Python version of LightGBM takes extremely high training time and I wanted to know if it's normal in Python LightGBM with large number of observations.

- Total observations: 75 million
- params : same for R and Python script
- features: nearly same for R and Python script
- Public LB: Almost same 
	- R version: 0.9690 
	- Python version: 0.9691 
- Modelling time:
	- R version: **24 mins** 
	- Python version: **278 mins**

Both script uses latest docker image and runs on 4 threads with maximum 16 GB RAM limit. 

Both scripts are publicly shared on Kaggle and below is the log details (including memory usage) for reproducibility. 

log of R script: https://www.kaggle.com/pranav84/single-lightgbm-in-r-with-75-mln-rows-lb-0-9690?scriptVersionId=2951218/log
log of Python script: https://www.kaggle.com/aharless/try-pranav-s-r-lgbm-in-python?scriptVersionId=2973530/log

Can someone please provide general feedback/thoughts on why Python version of LightGBM takes takes extremely high training time compared to R version. 

Thanks a lot.

**Comment by Laurae2** [comment]
@pranavpandya84 There should be no reason for Python to take that much time, as the underlying code is identical to R. Please use the same dataset and run without Kaggle kernels (on your machine).

Kaggle kernels are not reliable when it comes to CPU usage for benchmarking. Also, the Python kernel uses double the number of boosting iterations. Note also that the AUC differs significantly (features not being exactly identical can have a huge influence on the training time as with any boosting algorithm).

If you can't benchmark on your machine, please save the binary file and use it as input for R and Python for benchmarking (instead of loading from dataset in memory when creating the `lgb.Dataset`, load from  binary file).

Maybe @wxchan knows more about the data transformations done in the Python side?

**Comment by andyharless** [comment]
The links are screwy, because of the way Kaggle's web site handles them.  The Python kernel to which @pranavpandya84 intends to refer (go to https://www.kaggle.com/aharless/try-pranav-s-r-lgbm-in-python, click on "Versions", click the link for "Version 5", and click on "Log") does the same number of boosting iterations as the R kernel to which he refers (go to https://www.kaggle.com/pranav84/single-lightgbm-in-r-with-75-mln-rows-lb-0-9690, click on "Versions", click the link for "Version 4", and click on "Log").  The only major difference is that the Python version specifies `nthread=4`.  Also, while the features should have the same values, some of them may have different variable types (smaller, if anything, in the Python version).  The Python version takes about 16700 second to do 500 rounds, while the R version takes about 1400 seconds.  It isn't a proper benchmark, but, man, that ia a huge difference.

**Comment by Laurae2** [comment]
Multiple potential reasons below, the most probable is AUC which is left on training set in Python and should be removed as it is computationally extreme (significantly slower than training the model).

I've seen Kaggle kernels being 10x slower depending on the time of the day, and kernels where updates causes some libraries to take 5x or even longer time to do model training.

As LightGBM is memory intensive, if another person uses the same host to run his/her kernel, then the computation speed will be drastically lower.

This is why it would be better for comparison if all the following conditions for comparison are met:

* Binary dataset is used (not guaranteed here)
* Same number of threads is used (not guaranteed here)
* Run computations in an isolated environment (not Kaggle kernels, they are unreliable for comparison) (not guaranteed here)
* Same metric is used with the same sets (not guaranteed here, you have training set and AUC on 75M observations explodes computation time vs @pranavpandya84 using only validation set, which will be approximately 100x faster to compute metrics per iteration)

Because currently there's no way to know if it's a user mistake, a LightGBM issue, a Kaggle kernel issue, or a combination of those.

I just tested on Bosch (1M observations x 969 features): CLI, R, and Python timings are nearly identical (+-5%), on 56 threads for 1000 iterations.

@andyharless Try to remove the AUC computation from the training. It has an absurd computation cost when the number of observations > 10M. Myself I do this mistakes sometimes and it explodes the training time when AUC is used

Comparison: Higgs 10M, when adding AUC on training:

* x4-x5 computation time
* 80ms per training AUC
* 20ms per iteration

**Comment by andyharless** [comment]
Thank you @Laurae2. I didn't realize computing AUC for the training data was so expensive, and I didn't notice that difference between the kernels.  That may be the main explanation.  I'll try it without the training AUC after the current run finishes.

**Comment by pranavpandya84** [comment]
@Laurae2 Thanks you so much for the detailed response.

**Comment by andyharless** [comment]
Yes, the training AUC evaluation does appear to have been the problem. With it, the last Python run timed out after 6 hours. Without it, it ran in less than an hour. Thank you again, @Laurae2.
